### PR TITLE
feat(engine): support expanding variables in more functions

### PR DESCRIPTION
**Pull request by rajatjindal**
This adds expanding environment variables in `WithNewFile`.  ref: #7951. If this looks OK, I will add similar changes to other API's in this same PR.

### CODE REVIEW

File: core/schema/container.go
Comment (jedevc) [inline core/schema/container.go:777]: Move this into `core.ContainerExecOpts`, that way we don't duplicate between the legacy/new version of this.

Comment (rajatjindal) [reply core/schema/container.go:769]: do we need to add this to useExecLegacy? I thought we don;t need to add new abilities to that function as its supported with only older version of engines?

Comment (rajatjindal) [reply core/schema/container.go]: nvm, I see this is useful even if we are not making change to useExecLegacy. Moved the option to `core.ContainerExecOpts`.

**Comment by rajatjindal** [comment]
Hi @jedevc I believe this PR is ready for review. thank you.

**Comment by rajatjindal** [comment]
I am not quite sure why these tests are failing on CI. I have tried them on my local setup and it seems to have worked fine. (random failures in multiple tries). 

e.g. 

$ dagger -m . call test specific --run=TestModule --race=true --parallel=16

fails on CI: https://github.com/dagger/dagger/actions/runs/10895023969/job/30232651758?pr=8427
works on local: https://dagger.cloud/rajatjindal/traces/59b79095b18306af5d41551b5f9cdbc7

**Comment by rajatjindal** [comment]
The tests seems to be all green now 🎉

**Review by marcosnils — APPROVED** [review]

File: docs/docs-graphql/schema.graphqls
Comment (jedevc) [inline docs/docs-graphql/schema.graphqls:630]: `withMountedCache` needs an `ArgDoc` so we get the docs here as well.

File: .changes/unreleased/Added-20240916-192525.yaml
Comment (jedevc) [inline .changes/unreleased/Added-20240916-192525.yaml:2]:
```suggestion
body: Allow expanding environment variables in more `Container` functions.
```

File: core/schema/container.go
Comment (jedevc) [inline core/schema/container.go:1662]: nit: typo

File: core/schema/container.go
Comment (jedevc) [inline core/schema/container.go:88]: nit: formatting.

You can just use backtick strings for all of this.

File: .changes/unreleased/Added-20240916-192525.yaml
Comment (rajatjindal) [reply .changes/unreleased/Added-20240916-192525.yaml:2]: 👍

File: core/schema/container.go
Comment (rajatjindal) [reply core/schema/container.go:1662]: whoops. thank you. 👍

File: core/schema/container.go
Comment (rajatjindal) [reply core/schema/container.go:88]: the doc comment itself has backticks, so maybe we can keep it this way? (that way it also stays consistent with what we have in ExpandEnvVariable)

File: docs/docs-graphql/schema.graphqls
Comment (rajatjindal) [reply docs/docs-graphql/schema.graphqls:630]: Thank you very much for catching this 👍. fixed.

**Comment by rajatjindal** [comment]
Typescript test failed in CI - https://dagger.cloud/dagger/traces/cf2967cbf23f257e2c1e860f96cbba49?span=c73ef26b74be4ea3 

but works when running locally: https://dagger.cloud/rajatjindal/traces/5d4dc3da461f9794457d701c69f6d8d1

**Comment by rajatjindal** [comment]
all the tests are green now (with the in-flight fix for fixing go mod cache issue). I have now removed that commit, and I think this PR is ready for merge as well (pending one feedback from @jedevc). 

Thanks

File: core/schema/container.go
Comment (rajatjindal) [reply core/schema/container.go]: Hi @jedevc, I have merged the 2nd and 3rd line into a single line. The first line includes backticks in the string itself, so I don't think we can include complete line into single backtick'ed string. 

Kindly let me know if you have any alternate recommendation for handling this.

**Review by TomChv — APPROVED** [review]
LGTM to merge once the CI is fully green

I see this error with PHP, I wonder why it's doing that: 

```
error: parse selections: parse field "withExec": Container.withExec has no such argument: "expand"
/src/sdk/php/vendor/carnage/php-graphql-client/src/Results.php:54
/src/sdk/php/vendor/carnage/php-graphql-client/src/Client.php:160
/src/sdk/php/vendor/carnage/php-graphql-client/src/Client.php:115
/src/sdk/php/src/Client/AbstractClient.php:40
/src/sdk/php/src/Client/AbstractObject.php:26
/src/sdk/php/generated/Container.php:342
/src/sdk/php/tests/Integration/ClientTest.php:61
```

Will ping the PHP maintainer for investigation. Holding the merge till it's fixed

**Comment by rajatjindal** [comment]
Hi @TomChv 

Thank you for the review. 

> LGTM to merge once the CI is fully green

the PHP tests are expected to fail because of them not using dev engine for tests. other than that CI looks green. some context here: https://github.com/dagger/dagger/pull/8427#issuecomment-2354429740

**Comment by jedevc** [comment]
@rajatjindal I'll take a look at fixing the php ci today.

File: sdk/php/dev/src/PhpSdkDev.php
Comment (charjr) [inline sdk/php/dev/src/PhpSdkDev.php:28]:
```suggestion
    #[DaggerFunction]
    public function __construct(
        private ?Container $baseContainer = null,
    ) {
    }
```

This would assign it as a private property that can only be null or a Container.

PHP variables are usually in `camelCase` so `$baseContainer` is preferable.

Keep the property `private` unless we actually need access to it outside of the class.

File: sdk/php/dev/src/PhpSdkDev.php
Comment (charjr) [inline sdk/php/dev/src/PhpSdkDev.php:100]:
```suggestion
        return $this->baseContainer ?? dag()
```

Since this is how you're getting the base:
https://github.com/dagger/dagger/blob/fca16c064545c9e2f8c1d8c9e99bb1053af29bbc/.dagger/sdk_php.go#L75-L79

I don't think you'll need to mount the directory, as that function already does it:

https://github.com/dagger/dagger/blob/fca16c064545c9e2f8c1d8c9e99bb1053af29bbc/.dagger/sdk_php.go#L179-L193

The `$this->baseContainer ?? other thing` is shorthand for: "use `$this->baseContainer`, but if it's null, use `other thing`. If you want to understand it more [check here](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op)

**Review by charjr — CHANGES_REQUESTED** [review]
I think your code should do the job. But there's a few changes that could be done

**Comment by rajatjindal** [comment]
oh well, the CI is green now. :)

File: sdk/php/dev/src/PhpSdkDev.php
Comment (rajatjindal) [reply sdk/php/dev/src/PhpSdkDev.php:29]: aha, that explains why I had to use a different variable for making it work. thank you for the insight. I'll push the change.

File: sdk/php/dev/src/PhpSdkDev.php
Comment (jedevc) [inline sdk/php/dev/src/PhpSdkDev.php:38]: We can just remove this option, the default is `false`.

File: core/schema/container.go
Comment (jedevc) [inline core/schema/container.go:1187]: This snippet is very verbose and is repeated a lot. Can we have `expandPath(ctx, parent, expand, path)`?

**Review by charjr — APPROVED** [review]
I approve on the changes to `sdk/php/src/PhpSdkDev.php`

**Review by jedevc — COMMENTED** [review]
Just going through the list of methods in `Container`, noticed some fields that should probably also have this, for consistency:
- `withoutMount`
- `withUnixSocket`/`withoutUnixSocket`
- `withWorkdir`
- `withMountedSecret`

**Comment by shykes** [comment]
I'm testing it and a few things feel off:

## Usage message

```console
$ dagger call -m github.com/dagger/dagger@pull/8427/head dev with-exec --args=sh,-c,'dagger core container with-new-file --help'
[...]
      --expand ${VAR}     Replace ${VAR} or `$VAR` in the value of path
                          according to the current environment variables
                          defined in the container (e.g. "/$VAR/foo.txt").
```

The usage message doesn't seem to match reality: when I actually call `with-new-file --expand` it behaves like a boolean. So I assume a typo in the usage message?

## Expand not working

```console
dagger call -m github.com/dagger/dagger@pull/8427/head \
 dev \
 with-exec \
 --args=sh,-c,'dagger core container with-env-variable --name=FOO --value=/bar with-new-file --contents=hello --path=${FOO}.txt --expand rootfs glob --pattern=**' \
 stdout
```

Expected output: `bar.txt`
Actual output: `.txt`

## Expand always enabled?

```console
dagger call -m github.com/dagger/dagger@pull/8427/head \
 dev \
 with-exec \
 --args=sh,-c,'dagger core container with-env-variable --name=FOO --value=/bar with-new-file --contents=hello --path=${FOO}.txt --expand=false rootfs glob --pattern=**' \
 stdout
```

Expected output: `${FOO}.txt`
Actual output: `.txt`

**Comment by rajatjindal** [comment]
Hi @shykes, thanks for trying these changes out. 

> Expand always enabled?
dagger call -m github.com/dagger/dagger@pull/8427/head \
 dev \
 with-exec \
 --args=sh,-c,'dagger core container with-env-variable --name=FOO --value=/bar with-new-file --contents=hello --path=${FOO}.txt --expand=false rootfs glob --pattern=**' \
 stdout
Expected output: ${FOO}.txt
Actual output: .txt

we need to shell escape the dollar sign in `${FOO}.txt`, as otherwise it reaches the `WithNewFile` fn as `.txt`. e.g. refer to below screenshot from the trace:

Screenshot 2024-09-27 at 6 49 42 AM


and after escape:

Screenshot 2024-09-27 at 6 50 47 AM

also

> $ dagger call -m github.com/dagger/dagger@pull/8427/head dev with-exec --args=sh,-c,'dagger core container with-new-file --help'
[...]
      --expand ${VAR}     Replace ${VAR} or `$VAR` in the value of path
                          according to the current environment variables
                          defined in the container (e.g. "/$VAR/foo.txt").

I will check how this is rendered. the usage msg in the code looks ok.

**Comment by rajatjindal** [comment]
> > $ dagger call -m github.com/dagger/dagger@pull/8427/head dev with-exec --args=sh,-c,'dagger core container with-new-file --help'
[...]
--expand ${VAR} Replace ${VAR} or $VAR in the value of path
according to the current environment variables
defined in the container (e.g. "/$VAR/foo.txt").

> I will check how this is rendered. the usage msg in the code looks ok.

**Review by jedevc — APPROVED** [review]

**Comment by jedevc** [comment]
Thanks @rajatjindal, let's get this in :tada:

**Comment by jedevc** [comment]
Failure is a known flake, let's merge :tada:

**Pull request merged**
